pkg/streamingpromql/operators/functions/sort_by_label.go

```diff
@@ -0,0 +1,152 @@
+		memoryConsumptionTracker: memoryConsumptionTracker,
+		expressionPosition:       expressionPosition,
+	}
+}
+
+func (s *SortByLabel) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadata, error) {
+	innerMetadata, err := s.inner.SeriesMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Store the original indexes of each piece of metadata and then sort them alongside
+	// the metadata. This allows us to return series in sorted order when requesting them
+	// from the inner operator by their original index.
+	indexes := make([]int, len(innerMetadata))
+	for i := range indexes {
+		indexes[i] = i
+	}
+
+	sort.Sort(&sortableMetadata{
+		metadata:   innerMetadata,
+		indexes:    indexes,
+		labels:     s.labels,
+		descending: s.descending,
+	})
+
+	s.originalIndexes = indexes
+	s.buffer = operators.NewInstantVectorOperatorBuffer(s.inner, nil, len(innerMetadata), s.memoryConsumptionTracker)
+
+	return innerMetadata, nil
+}
+
+type sortableMetadata struct {
+	metadata   []types.SeriesMetadata
+	indexes    []int
+	labels     []string
+	descending bool
+}
+
+func (s *sortableMetadata) Len() int {
+	return len(s.metadata)
+}
+
+func (s *sortableMetadata) Swap(i, j int) {
+	s.metadata[i], s.metadata[j] = s.metadata[j], s.metadata[i]
+	s.indexes[i], s.indexes[j] = s.indexes[j], s.indexes[i]
+}
+
+func (s *sortableMetadata) Less(i, j int) bool {
+	res := s.lessInner(i, j)
+	if s.descending {
+		return !res
+	}
+	return res
+}
+
+func (s *sortableMetadata) lessInner(i, j int) bool {
+	a := s.metadata[i]
+	b := s.metadata[j]
+	for _, lbl := range s.labels {
+		val1 := a.Labels.Get(lbl)
+		val2 := b.Labels.Get(lbl)
+		if val1 == val2 {
+			continue
+		}
+
+		if natsort.Compare(val1, val2) {
+			return true
+		}
+
+		return false
+	}
+
+	// If all labels provided as arguments were equal, sort by the full label set. This ensures a
+	// consistent ordering and matches the behavior of the Prometheus engine.
+	return labels.Compare(a.Labels, b.Labels) < 0
+}
+
+func (s *SortByLabel) NextSeries(ctx context.Context) (types.InstantVectorSeriesData, error) {
+	if s.seriesIndex >= len(s.originalIndexes) {
+		return types.InstantVectorSeriesData{}, types.EOS
+	}
+
+	// Get the original position in the stream of series so that we can request it from
+	// the buffer which buffers series from the inner operator while they aren't needed.
+	originalIndex := s.originalIndexes[s.seriesIndex]
+
+	data, err := s.buffer.GetSeries(ctx, []int{originalIndex})
+	if err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
+
+	s.seriesIndex++
+	return data[0], nil
+}
+
+func (s *SortByLabel) ExpressionPosition() posrange.PositionRange {
+	return s.expressionPosition
+}
+
+func (s *SortByLabel) Prepare(ctx context.Context, params *types.PrepareParams) error {
+	return s.inner.Prepare(ctx, params)
+}
+
+func (s *SortByLabel) Close() {
+	// Closing the buffer also closes its source operator which is our `inner`.
+	s.buffer.Close()
```

Comment:
If `SeriesMetadata` wasn't called for some reason (or it returned an error before `s.buffer` is set), then we'll leave `inner` unclosed. All operators should be safe to close multiple times, so calling `inner.Close` here should be safe to do.

Reply:
Done in d5ff77f3902f7a9c3ef6dc6eac2c58f5f23daefe. It would correct to put the call to `s.inner.Close()` in the `else` statement but I'm not sure it's needed since we can call `.Close()` twice without issue.